Rewrite Osram Lightify component

OleksandrBerchenko opened the pull request:
## Description:

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation:** home-assistant/home-assistant.io#8621

Original code had multiple flaws: it was not performance efficient, used throttling incorrectly, implemented status of light groups in a confusing way, didn't take into account actual supported features of each device and lacked many functional features.  As a result, the module was re-implemented literally from scratch.  At the same time I did my best to keep the new code 100% compatible with the old version not to break existing installations (see also `Known limitations` below).

Summary of changes:

1. Upgrade underlying `python-lightify` library from `1.0.6.1` to `1.0.7.1`.  It was recently updated by me as well.  The changes include:

    - Dramatically improve performance by avoiding redundant queries and applying additional optimizations during parsing results
    - Add support for scenes
    - Set properties that don't exist for a particular device type to reasonable defaults
    - Expose supported features and ignore unsupported actions
    - Expose additional properties like firmware version and whether the lights are reachable or not
    - Add new device types and improve handling of unknown device types
    - Add built-in throttling mechanism
    - Additional refactoring

2. Further improve performance on Home Assistant level:

    - Use different frequency for updates of light statuses (need to be near real-time) and group configurations (for most users may be updated once per hour or even less frequently)
    - Fix throttling.  Existing throttling was simply broken: update function was always called with `no_throttle=True`.  Now built-in throttling of the underlying library is used.  It's more lightweight and also correctly handles errors: if some call failed, it doesn't throttle the next one.
    - Cache calculated values when possible to avoid unneeded recalculations

3. Add additional configuration values to allow users to control throttling and which types of devices (like swithes or sensors) should be loaded or not.

4. Add support for scenes.  If a group has associated scenes, they will be imported as light effects and visible in `Effect` dropdown on UI.  Also they may be applied using `light.turn_on` service.

5. Expose device availability and new properties provided by the underlying library.

6. Take into account supported features of each device (don't treat all devices as RGBW bulbs).

7. Correctly report statuses of light groups.

8. Other minor fixes and improvements.


Known limitations:

1. The module is still not able to distinguish "color" and "tunable white" modes of RGBW bulbs and to get battery levels of switches (that seems not possible using the local protocol).

2. Support for motion and contact sensors is not implemented.  It's definitely possible, but I don't own any of them to do an accurate fix and test it properly.  See also https://github.com/tfriedel/python-lightify/issues/9.  As a workaround, the module exposes a list of all raw sensor values and users may still use them for automations.  Please note that currently `osramlightify` is a pure `light` module, but making motion and contact sensors first-class citizens will require rewriting it to a custom component like `hue` or `xiaomi_aqara`.

3. The module does its best to support unknown devices (many 3rd party manufacturers start to produce Lightify-compatible bulbs), but minor issues like incorrect temperature ranges are still possible.  The module exposes original `device type id`, so now it should be easier to report bugs for specific devices.

4. `unique_id` for Lightify groups is calculated in a wrong way, but I have decided not to fix that not to break existing installations.  When the module is re-written to a custom component (see p. 2 above), it should be a right time to fix `unique_id` issue as well.

## Example entry for `configuration.yaml`:

```yaml
light:
  - platform: osramlightify
    scan_interval: 1
    host: 192.168.1.50
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

OleksandrBerchenko committed:
Completely rework Osram Lightify component

OleksandrBerchenko committed:
1. Expose raw sensor values, 2. refactor device specific attributes.

OleksandrBerchenko committed:
Bump lightify to 1.0.7.0

OleksandrBerchenko committed:
Update logging

OleksandrBerchenko committed:
Update python-lightify version to 1.0.7.1

OleksandrBerchenko committed:
Remove unneeded config.get statements

Kane610 commented:
This is an unnecessary large PR, combining new features and refactoring at the same time. You should split this up to smaller portions.

OleksandrBerchenko commented:
@Kane610 Your concern is valid, but please note the module was re-implemented from scratch (I used a word "refactoring" incorrectly).  I have updated my description above:

"Original code had multiple flaws: it was not performance efficient, used throttling incorrectly, implemented status of light groups in a confusing way, didn't take into account actual supported features of each device and lacked many functional features. As a result, the module was re-implemented literally from scratch. At the same time I did my best to keep the new code 100% compatible with the old version not to break existing installations (see also `Known limitations` below)."

Does it make sense?  Honestly I don't see an easy way to logically split it and I doubt that could help to compare the differences (because all the code is new, except for the general structure, inherited from the parent class).  I could artificially extract some biggest features like scenes just to minimize a total number of lines, but I'm not sure it's a good idea (and would require a lot of additional testing).

What do you think?

Thanks!

SukramJ commented:
Hello,
i saw you PR and i am very exited to get the rewritten component.
There is still something in it that disturbs me:
The unique_id of the group is a collection of the group members. So everytime the group members change a new light is created in HA.

A Fixture might be to use the group name and something like SHA1 (in this case security doesn't matter) to create a hash:
Livingroom ->4db1d0d44ea4a7e4036f08b60c1f49c4cc8bd0cf
Now add a prefix like you componentname and the result should be unique enough.
osramlightfy_4db1d0d44ea4a7e4036f08b60c1f49c4cc8bd0cf

I don't know the python implementation, but normally the hash should always be the same.

BR
Markus

OleksandrBerchenko commented:
@SukramJ Your concern is 100% valid, but as I mentioned in the description:

"4\. `unique_id` for Lightify groups is calculated in a wrong way, but I have decided not to fix that not to break existing installations.  When the module is re-written to a custom component (see p. 2 above), it should be a right time to fix `unique_id` issue as well."

I could fix that, but that would trigger new lights instead of old groups in each of existing HA installations.  I'm not 100% sure it is worth it.  I hope sooner or later this component will be extended to include not only lights, but also sensors (like `xiaomi_aqara`).  It's a backward incompatible change by itself, so probably it makes sense to apply all breaking changes at the same time.

Thanks!

SukramJ commented:
I installed it as custom_component and it works really good. 
Especially that light or plugs show only features they support is fine.

Please remove the debug output from the library.

BR
Markus

OleksandrBerchenko commented:
@SukramJ Are you referencing this line?

```
bridge = lightify.Lightify(host, log_level=logging.DEBUG)
```

It should not set the actual logging level, that code should just allow you to set DEBUG level when you need it like:

```
logger:
  default: info
  logs:
    lightify: debug
```

Without passing `log_level=logging.DEBUG` you are not able to set anything below INFO for the `lightify` module.

Do I miss anything?

Thanks!

SukramJ commented:
I haven't set any debug level in my config.
But my Output is full of debug stuff.
```
2019-02-21 17:54:43 DEBUG (SyncWorker_9) [lightify] Received "2 b'2903'"
2019-02-21 17:54:43 DEBUG (SyncWorker_9) [lightify] Length:   809
2019-02-21 17:54:43 DEBUG (SyncWorker_9) [lightify] Expected: 809
```
Hundrets of these lines.

After setting this:
```
logger:
  default: info
```
... the debug stuff is gone.

OleksandrBerchenko commented:
@SukramJ According to https://www.home-assistant.io/components/logger/:

```
default
    (log_level) (Optional) Default log level.
    Default value: debug
```

So, probably it's expected that you see all debug messages.  Did you see them only for this module or for the whole HA?

Thanks!

SukramJ commented:
I see them only for `lightify`.

BTW: This is a new installation with pip.

OleksandrBerchenko commented:
@SukramJ Ok, I will double check.  Thanks for the feedback!

OleksandrBerchenko commented:
@SukramJ I can't reproduce your issue with logging:

1. If `logger` is completely commented in `configuration.yaml`, I see only WARNING+ messages for all components.
2. If I have just:

    ```
    logger:
    ```

   I see DEBUG+ messages for all components (according to documentation).

3. If I have:

    ```
    logger:
      logs:
        lightify: info
    ```

   I see DEBUG+ messages for all components, but only INFO+ for `lightify` (according to documentation).

4. If I have:

    ```
    logger:
      default: info
    ```

   I see INFO+ messages for all components, including `lightify` (according to documentation).

5. If I have:

    ```
    logger:
      default: info
      logs:
        lightify: debug
    ```

   I see INFO+ messages for all components, and DEBUG+ for `lightify` (again, according to documentation).

Do I miss anything?  I'm using the latest `0.88.1` version.

Thanks!

OleksandrBerchenko committed:
Fix bug: internal is_on state was not updated during turn on/off

SukramJ commented:
This is my latest test setup:
- Debian 9 netinst
- install pip3
- install homeassistant `python3 -m pip install homeassistant`
- default configuration.yml with added lightfy:
```
homeassistant:
  # Name of the location where Home Assistant is running
  name: Home
  # Location required to calculate the time the sun rises and sets
  latitude: 53.5389
  longitude: 10.128
  # Impacts weather/sunrise data (altitude above sea level in meters)
  elevation: 0
  # metric for Metric, imperial for Imperial
  unit_system: metric
  # Pick yours from here: http://en.wikipedia.org/wiki/List_of_tz_database_time_zones
  time_zone: Europe/Berlin
  # Customization file
  customize: !include customize.yaml

# Configure a default setup of Home Assistant (frontend, api, etc)
default_config:

# Show the introduction message on startup.
introduction:

# Uncomment this if you are using SSL/TLS, running in Docker container, etc.
# http:
#   base_url: example.duckdns.org:8123

# Sensors
sensor:
  # Weather prediction
  - platform: yr

# Text to speech
tts:
  - platform: google

group: !include groups.yaml
automation: !include automations.yaml
script: !include scripts.yaml

light:
  - platform: osramlightify
    host: xx.xx.xx.xx
    allow_lightify_nodes: true
    allow_lightify_groups: true
```
- start hass
- osram lightfy works without debug logging
- `mkdir -p custom_components/light`
- copy `osramlightify.py` to new directory
- restart hass
- **now debug logging is in console**

I did not update the library in this latest test, but in previous test i did.
Everything works fine in HA with new component.

Have you tested osramlightify.py as **custom_component**?
Is the debug code related to custom_component? 

Thats all i can say.

Thanks.

OleksandrBerchenko commented:
@SukramJ Well, I never tried installing `osramlightify` as a custom component.  Also I don't think anybody will need to do that after the changes are merged.

Possibly logging configuration works differently for custom components.

Thanks!

OleksandrBerchenko committed:
Fix logging level

OleksandrBerchenko commented:
@SukramJ Could you please test again?  Now logging should work correctly in all cases (thanks @tfriedel for the fix).

SukramJ commented:
Logging is OK now.

dshokouhi commented:
@OleksandrBerchenko there seems to be a lot in this PR as mentioned earlier.  I don't think there is anything wrong with a rewrite to fix things however you are also adding features at the same time.  If you can split up this PR by removing all new features and just rewriting the current code it will probably get merged in faster.  Then your subsequent PR's for adding each new feature will get merged faster.  I don't have merge rights but I am speaking from experience here.

OleksandrBerchenko committed:
Move file to a new location

OleksandrBerchenko commented:
@dshokouhi Thanks for the suggestions.  Ok, I will try to manage that separation.  Anyway, I already got a conflict that I need to fix.

OleksandrBerchenko commented:
Ok, I have closed this one and opened the first of the new series: https://github.com/home-assistant/home-assistant/pull/22184.
